File: src/api/ApiError.fs
[<AutoOpen>]
module internal ApiErrorCode

open System

type ApiErrorCode =
    | ResourceNotFound
    | InvalidRequestParameter
    | PreconditionFailed
    | InvalidHeader
    | ResourceAlreadyExists
    | InvalidRequestBody

    static member toString value =
        match value with
        | ResourceNotFound -> nameof ResourceNotFound
        | InvalidRequestParameter -> nameof InvalidRequestParameter
        | PreconditionFailed -> nameof PreconditionFailed
        | InvalidHeader -> nameof InvalidHeader
        | ResourceAlreadyExists -> nameof ResourceAlreadyExists
        | InvalidRequestBody -> nameof InvalidRequestBody

    static member fromString value =
        match value with
        | _ when String.Equals(value, nameof ResourceNotFound, StringComparison.OrdinalIgnoreCase) -> ResourceNotFound
        | _ when String.Equals(value, nameof InvalidRequestParameter, StringComparison.OrdinalIgnoreCase) -> InvalidRequestParameter
        | _ when String.Equals(value, nameof PreconditionFailed, StringComparison.OrdinalIgnoreCase) -> PreconditionFailed
        | _ when String.Equals(value, nameof InvalidHeader, StringComparison.OrdinalIgnoreCase) -> InvalidHeader
        | _ when String.Equals(value, nameof ResourceAlreadyExists, StringComparison.OrdinalIgnoreCase) -> ResourceAlreadyExists
        | _ when String.Equals(value, nameof InvalidRequestBody, StringComparison.OrdinalIgnoreCase) -> InvalidRequestBody
        | _ -> invalidOp ($"'{value}' is not a valid API error code.")

File: src/api/ApiError.cs
namespace api;

public abstract record ApiErrorCode
{
    public sealed record ResourceNotFound : ApiErrorCode
    {
        public static ResourceNotFound Instance { get; } = new();

        public override string ToString() => nameof(ResourceNotFound);
    }

    public sealed record InvalidRequestParameter : ApiErrorCode
    {
        public static InvalidRequestParameter Instance { get; } = new();

        public override string ToString() => nameof(InvalidRequestParameter);
    }

    public sealed record InvalidRequestBody : ApiErrorCode
    {
        public static InvalidRequestBody Instance { get; } = new();

        public override string ToString() => nameof(InvalidRequestBody);
    }

    public sealed record ResourceAlreadyExists : ApiErrorCode
    {
        public static ResourceAlreadyExists Instance { get; } = new();

        public override string ToString() => nameof(ResourceAlreadyExists);
    }
}